x/market/types/id.go

```diff
@@ -39,6 +41,12 @@ func (id OrderID) Equals(other OrderID) bool {
 
 // Validate method for OrderID and returns nil
 func (id OrderID) Validate() error {
+	if id.Owner.Empty() {
+		return sdkerrors.Wrap(sdkerrors.ErrInvalidAddress, "owner cannot be empty")
+	}
+	if id.DSeq == 0 && id.GSeq == 0 && id.OSeq == 0 {
```

Comment:
I believe you can use `return id.GroupID().Validate()`.  Same goes for the rest of them, I'll illustrate below.

Reply:
I can't bc `OSeq` is not defined for `GroupID`

Reply:
* `BidID = OrderID + Provider`; `Validate(BidID) = Validate(OrderID) + Validate(Provider)`
* `OrderID = GroupID + Oseq`; `Validate(OrderID) = Validate(GroupID) + Validate(Oseq)` 
* `GroupID = DeploymentID + Gseq`; `Validate(GroupID) = Validate(DeploymentID) + Validate(GSeq)`

Reply:
In other words, re-use the validations from the contained structures.  Unrolling at every level will lead to bugs, missed validations, and make refactoring more difficult.  There are [utilities for converting between the various IDs](https://github.com/ovrclk/akash/blob/9161a1abe6bf21223ced0b81a2f5167a02d012e2/x/market/types/id.go).

See the [`Equals()`](https://github.com/ovrclk/akash/blob/9161a1abe6bf21223ced0b81a2f5167a02d012e2/x/market/types/id.go#L66-L69) methods for an example.

Reply:
I respectfully disagree, not having any validation or the lack of testing leads to bugs. The same applies for private functions and vars, they don't provide any extra security guarantee if you define the public functions within the scope of the module.

Reply:
> not having any validation or the lack of testing leads to bugs

That's why we have a task to do those things.

Those things are, of course, completely separate from the topic of the benefits of abstraction, unless you count the benefits of re-using already-tested code (when `OrderID.Validate()` is tested, `BidID.Validate()` would use that tested code).

> The same applies for private functions and vars, they don't provide any extra security guarantee if you define the public functions within the scope of the module.

I don't really know what you mean by that last part, but keeping the surface area of a package to a minimum interface is for abstraction (internal implementation can change without affecting users of the package), isolation (prevents and/or minimizes paths to state mutation), and yes, security (vars can be re-assigned).  This is software engineering 101 stuff.